### ISSUE TITLE
Consider adding new ConfigureAwait methods with additional TimeSpan as timeout or CancellationToken parameter. 

### ISSUE THREAD

**Issue by zlatanov** [dotnet/corefx#42292]
Consider introducing new awaitables (for both Tasks and ValueTasks) that can time out (throw OperationCanceledException) when awaited:

``` csharp
TimeoutConfiguredAwaitable<TResult> ConfigureAwait<TResult>( 
   bool continueOnCapturedContext, TimeSpan timeout);

TimeoutConfiguredAwaitable<TResult> ConfigureAwait<TResult>( 
    bool continueOnCapturedContext, CancellationToken cancellationToken);
```

Consider the following situations:

We have an expensive task that should be executed only once, but can be observed by many. We can use TaskCompletionSource<T> for this. However there is no built in mechanism for consumers to await the result for a given amount of time or some other condition (which can be linked easily to CancellationToken).

Another example is a singleton implementation where an application needs to have only one instance of a given type at any one time. The factory for this instance is asynchronous (for example establishing I/O connection). Consumers of this can await the factory to establish the connection (which may take longer than anticipated or even be impossible at the moment), but they need to be able to specify how much time they are willing to do that. Real world example in Azure AMQP library: https://github.com/Azure/azure-amqp/blob/c161ce1c5abd182a4751295f0bc2757481404c8a/src/Fx/Singleton.cs#L113. Here the implementation is incorrect and is not honoring the timeout passed in when awaiting the factory task.

Here is an old popular question from stackoverflow: https://stackoverflow.com/questions/4238345/asynchronously-wait-for-taskt-to-complete-with-timeout.

Example of how the usage of this would look like:
``` csharp
await _tcs.Task.ConfigureAwait(timeout: TimeSpan.FromSeconds(5));

// Or another example if I'm using it inside ASP.NET Core Request
await _tcs.Task.ConfigureAwait(cancellationToken: HttpContext.RequestAborted);
```

Here is an example of how this can be implemented without having access to the internals of the Tasks - https://gist.github.com/zlatanov/8ed79f1897e5b737307532b10062ec39. I am hoping that there might be a better implementation and support if done here internally.

**Comment by Clockwork-Muse** [comment]
Current workaround:
```csharp
await Task.WhenAny(whateverInitializationTask, new Task(() => {}, yourCancellationToken));
```
The other task will be created with `TaskStatus.Created`, and if you never start it it will not complete.  
If necessary, you can chain `ContinueWith` calls on the dummy task to do things like provide defaults on timeouts.

**Comment by tarekgh** [comment]
@zlatanov could you please format the request as described in https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md?

**Comment by sharwell** [comment]
Here's an example `WithTimeout` implementation:

https://github.com/microsoft/vs-threading/blob/09e774c49bc5bf56470455951d7dc3cd0c0a0920/src/Microsoft.VisualStudio.Threading/TplExtensions.cs#L63-L109

**Comment by tarekgh** [comment]
CC @stephentoub

**Comment by scalablecory** [comment]
People very rarely want to cancel/timeout the await itself rather than the actual async operation --- making this ease to use *correctly* is imperative.
